src/core/xfa/factory.js

```diff
@@ -45,9 +46,29 @@ class XFAFactory {
     return this.root && this.form;
   }
 
-  _createPages() {
+  _getPages() {
+    const iterator = this.form[$toDocument]();
+    return new Promise((resolve, reject) => {
+      const lambda = () => {
+        try {
+          const value = iterator.next();
+          if (value.done) {
+            resolve(value.value);
+          } else {
+            setTimeout(lambda, 0);
+          }
+        } catch (e) {
+          reject(e);
+        }
+      };
+      setTimeout(lambda, 0);
+    });
+  }
+
+  async _createPages() {
     try {
-      this.pages = this.form[$toHTML]();
+      this.pages = await this._getPages();
```

Comment:
*Given that testing so far seem positive, I'm adding another small comment here.*

We this new method, we now have `getPages`, `_createPages`, and `_getPages` which feels a bit confusing. How about either:
 - Given the small size of both of these methods, how about just inlining the new code directly here instead?
 - Or at least calling the new method e.g. `_createPagesHelper` instead, to reduce some naming-confusion here.